subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSrcBuildListenerFactory.java

```diff
@@ -17,68 +17,79 @@
+        return new Listener(buildSrcRootProjectConfiguration, instantiator, classpathTransformer);
     }
 
     /**
      * Inspects the build when configured, and adds the appropriate task to build the "main" `buildSrc` component.
      * On build completion, makes the runtime classpath of the main `buildSrc` component available.
      */
-    public static class Listener extends InternalBuildAdapter implements ModelConfigurationListener {
-        private FileCollection classpath;
+    public static class Listener extends InternalBuildAdapter implements EntryTaskSelector {
+        private Configuration classpathConfiguration;
         private ProjectState rootProjectState;
         private final Action<ProjectInternal> rootProjectConfiguration;
+        private final ScriptClassPathResolver resolver;
 
-        private Listener(Action<ProjectInternal> rootProjectConfiguration) {
+        private Listener(Action<ProjectInternal> rootProjectConfiguration, NamedObjectInstantiator instantiator, CachedClasspathTransformer classpathTransformer) {
             this.rootProjectConfiguration = rootProjectConfiguration;
+            this.resolver = new DefaultScriptClassPathResolver(Collections.emptyList(), instantiator, classpathTransformer);
         }
 
         @Override
         public void projectsLoaded(Gradle gradle) {
-            rootProjectConfiguration.execute((ProjectInternal) gradle.getRootProject());
+            GradleInternal gradleInternal = (GradleInternal) gradle;
+            // Run only those tasks scheduled by this selector and not the default tasks
+            gradleInternal.getStartParameter().setTaskRequests(Collections.emptyList());
+            ProjectInternal rootProject = gradleInternal.getRootProject();
+            rootProjectState = rootProject.getOwner();
+            rootProjectConfiguration.execute(rootProject);
         }
 
         @Override
-        public void onConfigure(GradleInternal gradle) {
-            final BuildableJavaComponent mainComponent = mainComponentOf(gradle);
-            gradle.getStartParameter().setTaskNames(mainComponent.getBuildTasks());
-            classpath = mainComponent.getRuntimeClasspath();
-            rootProjectState = gradle.getRootProject().getOwner();
-        }
-
-        public Collection<File> getRuntimeClasspath() {
-            return rootProjectState.fromMutableState(p -> classpath.getFiles());
+        public void applyTasksTo(Context context, ExecutionPlan plan) {
+            rootProjectState.applyToMutableState(rootProject -> {
+                classpathConfiguration = rootProject.getConfigurations().create("buildScriptClasspath");
+                classpathConfiguration.setCanBeConsumed(false);
+                resolver.prepareClassPath(classpathConfiguration, rootProject.getDependencies());
+                classpathConfiguration.getDependencies().add(rootProject.getDependencies().create(rootProject));
```

Comment:
If I'm reading this correctly... is rootProject here the root of _buildSrc_? 

Isn't this still ultimately the runtimeClasspath? What's the purpose of buildScriptClasspath?

Reply:
There are a couple of ways `buildScriptClasspath` is different to `runtimeClasspath` (the configuration). The first is that we want it to contain the outputs of `buildSrc` + its runtime dependencies.

The second is that I think we should resolve it as a project dependency on `buildSrc` in the same way that the runtime classpath for a plugin from an included build is resolved.

The idea is that this `Configuration` is just an intermediate step, and later we return a `dependency` out of here instead of a classpath and the code that wires in the plugins from included builds can take care of adding in `buildSrc` as well. I didn't get quite that far on the trip plus there are a couple of potential issues with this, as you point out.

Now that I think about it, this should probably be a detached configuration, except I guess we'll lose the progress logging if we do that.

Reply:
I think it's better to deal with a new named configuration than introduce a detached configuration that'll show up in every build with buildSrc on build scans.